### PR TITLE
Move the logic to models.py and migrate the script to argparse

### PR DESCRIPTION
**Pull request by pypingou**
There are two commits in this PR, the first one move the logic for moving an update from pending to stable directly, out of the approve_testing script and into ``bodhi/server/models.py`` as a method of the ``Update`` objects.
This will allow to re-use this heuristic in other places of bodhi without duplicating the code.

The second commit move the approve_testing script to argparse to make it easier to support a ``--debug`` option allowing to change the logging level from ``ERROR`` to ``INFO``, thus giving the user running the script more info about what is going on.

### CODE REVIEW

File: bodhi/server/models.py
Comment (Zlopez) [inline bodhi/server/models.py:3661]: Do you need the session to be passed as argument?

**Review by Zlopez — COMMENTED** [review]
Only one comment, otherwise looks good

File: bodhi/server/models.py
Comment (nphilipp) [reply bodhi/server/models.py:3661]: It's how methods working on a database session are usually implemented in Bodhi.

File: bodhi/server/models.py
Comment (nphilipp) [inline bodhi/server/models.py:3663]: Rather than mentioning the types in the docstring, would you add type hints? Our Python baseline is 3.6 and (AIUI) we want new code to have type hints rather than the same info in docstrings (but haven't gotten around to documenting this, ouch).

File: bodhi/server/models.py
Comment (nphilipp) [inline bodhi/server/models.py]: I don't think the session should be committed here, rather than where the method is used.

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [inline bodhi/tests/server/scripts/test_approve_testing.py:202]: To cope with debugging output added elsewhere in `models`, I'd prefer something like this:

```
self.assertIn(
    call.info(f'{update.alias} now meets testing requirements'),
    log.mock_calls)
```

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [inline bodhi/tests/server/scripts/test_approve_testing.py]: Similar here:

```
for msg_call in (call.debug(f'Set alias for bodhi2-2.0-1.fc17 to {update2.alias}'),
            call.info(f'{update2.alias} now meets testing requirements'),
            call.info(f'{update.alias} now meets testing requirements')):
    self.assertIn(msg_call, log.mock_calls)
```

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [inline bodhi/tests/server/scripts/test_approve_testing.py]: "... 1 argument ..."?

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [inline bodhi/tests/server/scripts/test_approve_testing.py:165]: This fails in local tests, it returns 30, i.e. `log.WARNING`. Maybe some other test messes around with logging levels?

I'm not sure what the best way to fix this would be:

* Either intercept/mock out the methods of `models.log` that can set the log level so it isn't actually globally changed in tests.
* Or, instead of using `mock.patch` to get hold of the log object, use the `caplog` fixture of pytest. See  https://docs.pytest.org/en/latest/logging.html#caplog-fixture
* Something else?

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [inline bodhi/tests/server/scripts/test_approve_testing.py]: This fails in local tests because I call pytest as `pytest` instead of `py.test-3`. Either take the actual executable name into account here, or perhaps use a regex match?

**Review by Zlopez — APPROVED** [review]

File: bodhi/server/scripts/approve_testing.py
Comment (nphilipp) [inline bodhi/server/scripts/approve_testing.py:61]: This feels odd, that `--debug` sets the log level to `INFO`. Should this be a `--verbose` flag instead?

**Review by nphilipp — CHANGES_REQUESTED** [review]
I spotted some changes in the second commit fixing stuff from the first (e.g. `call.debug()` vs. `call.info()`), not sure how to flag this appropriately here. :)

File: bodhi/server/scripts/approve_testing.py
Comment (pypingou) [reply bodhi/server/scripts/approve_testing.py:61]: wfm

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (pypingou) [reply bodhi/tests/server/scripts/test_approve_testing.py]: Already have the fix for this :)

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (pypingou) [reply bodhi/tests/server/scripts/test_approve_testing.py]: Good catch

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (pypingou) [reply bodhi/tests/server/scripts/test_approve_testing.py:165]: I have the same problem locally so I suspect something is messing up w/ logs somewhere :(

File: bodhi/server/models.py
Comment (pypingou) [reply bodhi/server/models.py:3663]: If we don't commit, there is no need to pass the db object so that makes this comment moot :)

File: bodhi/server/models.py
Comment (pypingou) [reply bodhi/server/models.py]: Simplifies things :)

File: bodhi/server/models.py
Comment (pypingou) [reply bodhi/server/models.py:3663]: Actually, the db object is needed in a few places, so nm my last comment (just the Raises is to be removed)

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (pypingou) [reply bodhi/tests/server/scripts/test_approve_testing.py:202]: We would have to have this debug info be added in the code path used in this part of the code, so do you still want this?

File: bodhi/server/models.py
Comment (nphilipp) [reply bodhi/server/models.py:3663]: > Actually, the db object is needed in a few places, so nm my last comment (just the Raises is to be removed)

Do you know if implicit flushes could still raise `SQLAlchemyError`?

File: bodhi/server/models.py
Comment (nphilipp) [reply bodhi/server/models.py]: Heh. It's just that we don't ever commit to the DB elsewhere in the model classes. :)

File: bodhi/server/models.py
Comment (Zlopez) [reply bodhi/server/models.py:3663]: Yes, they can. I was doing something like this in Anitya tests.

Comment (pypingou) [reply bodhi/server/models.py:3663]: ``.flush()`` will raise any exception that ``.commit()`` would

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [reply bodhi/tests/server/scripts/test_approve_testing.py:202]: We're calling `self.comment()` and `self.set_request()` from within `Update.move_to_stable()` so I'd not totally rule out that some time down the road, someone adds debugging in some unrelated place which might break on this -- then -- unrelated test.

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [reply bodhi/tests/server/scripts/test_approve_testing.py]: 👍

File: bodhi/server/scripts/approve_testing.py
Comment (nphilipp) [reply bodhi/server/scripts/approve_testing.py:61]: Heh, if anybody is ever confused about this, I'll send them your way. :stuck_out_tongue:

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (pypingou) [reply bodhi/tests/server/scripts/test_approve_testing.py:202]: okido :)

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (nphilipp) [reply bodhi/tests/server/scripts/test_approve_testing.py:165]: Shall I give option 2 (pytest caplog fixture) a try?

File: bodhi/server/scripts/approve_testing.py
Comment (pypingou) [reply bodhi/server/scripts/approve_testing.py:61]: I've adjusted it! :D

File: bodhi/server/models.py
Comment (nphilipp) [reply bodhi/server/models.py:3663]: So I'd leave the `Raises: ...` block in because implicit flushes are a thing.

File: bodhi/tests/server/scripts/test_approve_testing.py
Comment (pypingou) [reply bodhi/tests/server/scripts/test_approve_testing.py:165]: Feel free to give it a try, but from my quick testing I couldn't make it work, if you've played with it before you may have better luck though :)

File: bodhi/server/models.py
Comment (pypingou) [reply bodhi/server/models.py:3663]: okido

File: bodhi/server/scripts/approve_testing.py
Comment (cverna) [inline bodhi/server/scripts/approve_testing.py:84]: I don't think we need this, this script is installed by setup.py and points to its main function (https://github.com/fedora-infra/bodhi/blob/develop/setup.py#L232)

File: bodhi/server/models.py
Comment (cverna) [inline bodhi/server/models.py:3656]: This is misleading since it will not actually move the update to stable, but request the update to be move to stable by the next compose. 
The effect is that bodhi will tag the koji build with the release `Pending stable tag`. For the rawhide use case we still need to find out how the update status will be updated to stable after the nightly compose.

Maybe we should keep the name as `approve_testing` or something in that spirit.

File: bodhi/server/models.py
Comment (cverna) [inline bodhi/server/models.py:3658]: See above comment :-), also the code does more than that in some case it will not request the update to be pushed to stable.

**Review by cverna — COMMENTED** [review]
Added a few comments, overall I am still not convince we need this change at this point of time. What I would do is change the frequency of the approve testing cron job to have it run every 10 min for example and let people play with that in staging.
Based on the feedbacks we should consider improving the experience or not :-).

Just by curiosity what is the need to have the build in the rawhide buildroot available immediately ? Is that for building dependent packages ? if that's the case would that not be covered by the multi build update ?

File: bodhi/server/scripts/approve_testing.py
Comment (pypingou) [reply bodhi/server/scripts/approve_testing.py:84]: It is present in ``skopeo_lite.py``, ``check_policies.py`` and ``sar.py`` but I'm fine with removing it, I just find it  convenient.

**Comment by pypingou** [comment]
> Just by curiosity what is the need to have the build in the rawhide buildroot available immediately ? Is that for building dependent packages ? if that's the case would that not be covered by the multi build update ?

One of the requirement is to not impact the current workflow in rawhide for package that do not opt-in. Waiting 10 minutes, it impacting the current workflow. People expect their package to land in rawhide in a few minutes after they have built it.

File: bodhi/server/models.py
Comment (pypingou) [reply bodhi/server/models.py:3656]: This makes me think that this is not working or going to work as we want.

If you check: ``koji -s https://koji.stg.fedoraproject.org/kojihub buildinfo python-arrow-0.12.1-10.fc31`` you will see that this build is still in ``f31-updates-testing`` while the build is 8 days old.

If the move to stable only happens at the compose time, since we are not doing compose in bodhi for rawhide, it means updates will be stuck in ``f31-updates-testing``, potentially, in a pending state in the UI, but nothing moves the update further :(

Comment (cverna) [reply bodhi/server/models.py:3656]: yes also the f31 release in staging is not configure properly, the `Pending stable tag` should be `f31-pending` and not `f31-updates-testing` so that the builds get signed by robosignatory and tagged with the `f31` tag.

I guess we need a consumer that listen to the end of compose messages, and set the updates status to `STABLE` if the compose was successful.

**Comment by cverna** [comment]
> > Just by curiosity what is the need to have the build in the rawhide buildroot available immediately ? Is that for building dependent packages ? if that's the case would that not be covered by the multi build update ?
> 
> One of the requirement is to not impact the current workflow in rawhide for package that do not opt-in. Waiting 10 minutes, it impacting the current workflow. People expect their package to land in rawhide in a few minutes after they have built it.

In my opinion this is the perfect example where getting early feedback from users is useful, we have a good enough solution which makes a slight impact to the current workflow instead of assuming that this is wrong we should ask users if they can live with that or not. If not we can improve the solution, if they can live with it then happy days we saved ourselves some work :-)

File: bodhi/server/scripts/approve_testing.py
Comment (cverna) [reply bodhi/server/scripts/approve_testing.py:84]: Yeah, I am just trying to be conservative and make sure we don't break things that nobody was aware of :stuck_out_tongue:

**Comment by pypingou** [comment]
> In my opinion this is the perfect example where getting early feedback from users is useful, we have a good enough solution which makes a slight impact to the current workflow instead of assuming that this is wrong we should ask users if they can live with that or not. If not we can improve the solution, if they can live with it then happy days we saved ourselves some work :-)

I agree but the only place where we can have early feedback is in staging. Once we turn this on in production, everyone will be impacted by this change.
And in staging things are still not working and asking for feedback there means asking people to do work in two places, one of which is thrown away (might still be worth doing, the first point in this sentence is what worries me most).

File: bodhi/server/models.py
Comment (pypingou) [reply bodhi/server/models.py:3656]: This means we need to figure out exactly which update went into every compose.
To avoid this work which is tedious and error-prone, we agreed that we would mark the updates as STABLE once they are available in the rawhide buildroot, just as it is today (once the build is in the buildroot, there is nothing else to consider for rawhide).

So the flow is:
- build in koji
- automatically create an update in bodhi in a PENDING state
- query greenwave to know if that update should be gated or not (did they opt-in?) if not, skip the next step
- wait for greenwave to announce its decision
- inform the user about greenwave's decision by a comment on the udpate
- if greenwave lets the update go through, push the update to the ``-pending`` tag and mark the update as STABLE.

Comment (cverna) [reply bodhi/server/models.py:3656]: > This means we need to figure out exactly which update went into every compose.
> To avoid this work which is tedious and error-prone, we agreed that we would mark the updates as STABLE once they are available in the rawhide buildroot, just as it is today (once the build is in the buildroot, there is nothing else to consider for rawhide).
> 
> So the flow is:
> 
>     * build in koji
> 
>     * automatically create an update in bodhi in a PENDING state
> 
>     * query greenwave to know if that update should be gated or not (did they opt-in?) if not, skip the next step
> 
>     * wait for greenwave to announce its decision
> 
>     * inform the user about greenwave's decision by a comment on the udpate
> 
>     * if greenwave lets the update go through, push the update to the `-pending` tag and mark the update as STABLE.

Ha ok yes that makes things simpler :-). So we just need something like that in the approve_testing script : 

```
if update.release.mandatory_days_in_testing == 0:
    update.status =  UpdateStatus.stable

```
after this line https://github.com/fedora-infra/bodhi/blob/develop/bodhi/server/scripts/approve_testing.py#L105.

We have to keep in mind that the build might not be in stable tho depending on the signing status. Also I think users are used to be have the build available in the repo when it is marked as stable in bodhi so that might get them confused. 

Maybe we should approach the packagers that have originally commented on the Change proposal and get there inputs on these decisions.

**Comment by nphilipp** [comment]
> Just by curiosity what is the need to have the build in the rawhide buildroot available immediately ? Is that for building dependent packages ? if that's the case would that not be covered by the multi build update ?

AIUI, the multi build update involves using a side tag, so this is, at the very least, a change of process for pkg maintainers, if not more manual work. This is anecdotal of course, but when I have my packager hat on the sequence of 1) build something in Rawhide 2) koji wait-repo 3) build something else that depends on it (lather, rinse, repeat) is a very common one. Everything that makes it longer or more tedious won't be appreciated. :wink:

> Maybe we should approach the packagers that have originally commented on the Change proposal and get there inputs on these decisions.

Is this a thread on devel list? If yes, from when (circa)?

**Comment by pypingou** [comment]
> Is this a thread on devel list? If yes, from when (circa)?

It was done a bit in emails after the original thread on the rawhide package gating change proposal and followed up by an IRC meeting where we gathered some more inputs leading to the refinement of the change proposal and it being merged back into one proposal instead of two.

**Comment by pypingou** [comment]
> > Just by curiosity what is the need to have the build in the rawhide buildroot available immediately ? Is that for building dependent packages ? if that's the case would that not be covered by the multi build update ?
> 
> AIUI, the multi build update involves using a side tag, so this is, at the very least, a change of process for pkg maintainers, if not more manual work. This is anecdotal of course, but when I have my packager hat on the sequence of 1) build something in Rawhide 2) koji wait-repo 3) build something else that depends on it (lather, rinse, repeat) is a very common one. Everything that makes it longer or more tedious won't be appreciated. wink

Agreed, especially since this is going to also impact the person not opting-in in gating. At this point, however, I'm ok with running the cron job every 5 minutes or 2 minutes even to see if we can get the flow to work in staging and hopefully we can then improve this in time for production.

File: bodhi/server/models.py
Comment (bowlofeggs) [reply bodhi/server/models.py:3656]: We should not pass the db session in new Bodhi code. I spent the past three years trying to get rid of that pattern. You can get the db session by instantiating ```bodhi.server.models.Session```.

File: bodhi/server/scripts/approve_testing.py
Comment (bowlofeggs) [reply bodhi/server/scripts/approve_testing.py:84]: Yeah let's not put this code - it's not easy to test and is also not necessary, and in fact also cannot be used.

File: bodhi/server/scripts/approve_testing.py
Comment (bowlofeggs) [inline bodhi/server/scripts/approve_testing.py:25]: Let's not use argparse - we use click everywhere else that we need a CLI lib.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
I don't think we need a ```--debug``` flag. We have config files to change logging levels - that's how they should be changed.

The PR message says we are moving the code for other uses (I suppose future uses?) - what are those?

**Comment by cverna** [comment]
Let's close this, we can reopen it, if we still think that this is needed

**Pull request closed without merging**
